Civi/Civioffice/Wopi/Validation/WopiProofValidator.php

```diff
@@ -0,0 +1,134 @@
+      return FALSE;
+    }
+
+    $expectedProof = $this->getExpectedProof($proofValidatorInput);
+    $rsa = $this->createRsa($proofKey);
+    $oldRsa = $this->createRsa($oldProofKey);
+
+    $proof = (string) base64_decode($proofValidatorInput->proof, TRUE);
+    $oldProof = (string) base64_decode($proofValidatorInput->oldProof, TRUE);
+
+    // The X-WOPI-Proof value using the current public key.
+    return $rsa->verify($expectedProof, $proof)
+      // The X-WOPI-ProofOld value using the current public key.
+      || $rsa->verify($expectedProof, $oldProof)
+      // The X-WOPI-Proof value using the old public key.
+      || $oldRsa->verify($expectedProof, $proof);
+  }
+
+  private function getExpectedProof(ProofValidatorInput $proofValidatorInput): string {
+    $accessToken = $proofValidatorInput->accessToken;
+    $url = strtoupper($proofValidatorInput->url);
+    // Timestamp as unsigned long long (64 bit, big endian byte order)
+    $timestamp = pack('J', $proofValidatorInput->timestamp);
+
+    return sprintf(
+      '%s%s%s%s%s%s',
+
+      // Length of access token as unsigned long (32 bit, big endian byte order)
+      pack('N', strlen($accessToken)),
+      $accessToken,
+
+      // Length of the full request URL as unsigned long (32 bit, big endian byte order)
+      pack('N', strlen($url)),
+      $url,
+
+      // Length of the WOPI timestamp as unsigned long (32 bit, big endian byte order)
+      pack('N', strlen($timestamp)),
+      $timestamp
+    );
+  }
+
+  /**
+   * Construct the RSA public key from modulus and exponent.
+   *
+   * @throws \InvalidArgumentException
+   */
+  private function createRsa(WopiProofKey $key): RSA {
+    $modulus = base64_decode($key->modulus, TRUE);
+    if (FALSE === $modulus) {
+      throw new \InvalidArgumentException('Invalid modulus in WOPI proof public key');
+    }
+
+    $exponent = base64_decode($key->exponent, TRUE);
+    if (FALSE === $exponent) {
+      throw new \InvalidArgumentException('Invalid exponent in WOPI proof public key');
+    }
+
+    $rsa = new RSA();
+    if (!$rsa->loadKey([
+      'e' => new BigInteger($exponent, 256),
+      'n' => new BigInteger($modulus, 256),
+    ])) {
+      throw new \InvalidArgumentException('Invalid WOPI proof public key');
+    }
+
+    $rsa->setSignatureMode(RSA::SIGNATURE_PKCS1);
+    $rsa->setHash('sha256');
+
+    return $rsa;
+  }
+
+  /**
+   * Verify WOPI timestamp and make sure that it was sent within the last 20 minutes.
+   */
+  private function verifyTimestamp(ProofValidatorInput $proofValidatorInput): bool {
+    $timestamp = $proofValidatorInput->timestamp;
+    if (!is_numeric($timestamp)) {
+      return FALSE;
+    }
+
+    // WOPI timestamps are the number of 100 nanosecond units passed since 1/1/0001.
+    $date = DotNetTimeConverter::toDateTime((int) $timestamp);
+
+    $timestampDiff = abs(\CRM_Utils_Time::time() - $date->getTimestamp());
+    if ($timestampDiff > 20 * 60) {
```

Comment:
Move the *20 minutes* to a constant?

Reply:
The 20 minutes is a requirement from the [specification](https://learn.microsoft.com/en-us/microsoft-365/cloud-storage-partner-program/online/scenarios/proofkeys). To me a constant like `MAX_PROOF_AGE` looks more like something that can be changed if someone thinks it makes sense than having the value directly in the code. That's why I'd prefer not to use a constant in this case.